src/App.tsx

```diff
@@ -1,29 +1,70 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { useSearchParams } from 'react-router-dom';
+
 import './App.css';
-import { getNumbers } from './utils';
+import { getNumbers, preparePageItems } from './utils';
+import { Pagination } from './components/Pagination';
 
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
 const items = getNumbers(1, 42).map(n => `Item ${n}`);
 
+const paginationOptions: number[] = [3, 5, 10, 20];
+
 export const App: React.FC = () => {
+  const [searchParams, setSearchParams] = useSearchParams();
+
+  const getSearchParam = (param: string, defaultValue: number) =>
+    Number(searchParams.get(param) ?? defaultValue);
+
+  const [perPage, setPerPage] = useState(
+    getSearchParam('perPage', paginationOptions[1]),
+  );
+
+  const [currentPage, setCurrentPage] = useState(getSearchParam('page', 1));
+
+  const showItems = preparePageItems(items, perPage, currentPage);
+  const pageHeader = `Page ${currentPage} (items ${perPage * (currentPage - 1) + 1} - ${Math.min(perPage * currentPage, items.length)} of ${items.length})`;
+
+  const perPageSelectHandler = (
+    event: React.ChangeEvent<HTMLSelectElement>,
+  ) => {
+    setPerPage(Number(event.target.value));
+    setCurrentPage(1);
+  };
+
+  const pageChangeHandler = (page: number) => {
+    if (page !== currentPage) {
+      setCurrentPage(page);
+    }
+  };
+
+  useEffect(() => {
+    setSearchParams(`?page=${currentPage}&perPage=${perPage}`);
+  }, [currentPage, perPage, searchParams, setSearchParams]);
```

Comment:
Including `searchParams` and `setSearchParams` in the dependency array of `useEffect` is unnecessary because they are stable references provided by `useSearchParams`. This could lead to unnecessary re-renders. Consider removing them from the dependency array.